### PR TITLE
Allow hash in color schemes yaml files

### PR DESCRIPTION
**Pull request by belak**
The current phrasing says "should not" which isn't very clear and also doesn't explicitly disallow it. I'd like to propose that the hash can optionally be included. There are a number of different editor plugins which can highlight colors, but they usually require that they're preceded by a `#` so it corresponds with an html color code.

### CODE REVIEW

**Comment by belak** [comment]
This technically needs a version bump, so if this goes in, it may be combined with other changes.

**Comment by joshgoebel** [comment]
> The current phrasing says "should not" which isn't very clear and also doesn't explicitly disallow it.

I find it pretty clear.  How would you rephrase it to explicitly disallow?

> This technically needs a version bump because it's a breaking change

This would also break every single builder I'm pretty sure (which would need to know about this variant)  so I'm not sure how to push something like this thru without breaking the whole ecosystem...?  I think to even consider this there would have to be some sort of plan there?

**Comment by belak** [comment]
> I find it pretty clear. How would you rephrase it to explicitly disallow?

"MUST NOT" is often used in RFCs.

> This would also break every single builder I'm pretty sure (which would need to know about this variant) so I'm not sure how to push something like this thru without breaking the whole ecosystem...? I think to even consider this there would have to be some sort of solid plan there?

I maintain one of the most up-to-date builders. Generally, if I push through a spec change, I release an updated builder as well.

That being said, I forgot I opened this PR. There are upsides and downsides to both approaches - allowing `#` lets packages which highlight HTML colors work, but disallowing them makes it harder for someone to accidentally mess up and release it with empty strings for colors and comments at the end of the line.

I do think it makes sense to just pick one - either explicitly allow ("may optionally"), or explicitly disallow ("must not") would be better than what's there right now.

**Comment by joshgoebel** [comment]
> I do think it makes sense to just pick one - either explicitly allow ("may optionally"), or explicitly disallow ("must not") would be better than what's there right now.

I'd be fine with "MUST NOT" - that is certainly stronger. :-)  And if we were starting over then requiring the `#` by default I think (the ubiquity of CSS) makes sense.  Just imagining a world where everyone did it differently and it wasn't pretty.

**Comment by belak** [comment]
We've also been looking at batching a number of these changes if/when we migrate some of these repos to a github org. If we wanted to explicitly allow `#`, doing it then may make sense. Anyway, I haven't made a decision one way or another. My base16-builder-go implementation allows the # but doesn't require it, which I think is in line with what the spec currently says. The clarification is mainly because we've had a number of people submit themes where they forgot to quote the string and so they accidentally submitted an empty scheme. This will go away with the 0.10.0 spec proposal, because all themes will be in one repo and we'll have better control over them.

Sorry, I'm rambling a bit.

Like I originally said, I forgot I opened this PR and it hasn't come up much since I opened it which may mean it's not worth clarifying.

**Comment by joshgoebel** [comment]
I guess I was just mostly worried about leaving older builders in the dust, but I guess that's what we already seem to do as a policy - so maybe that's just the way it is? :-)  I've been updating the Node builder to the latest spec.

**Comment by belak** [comment]
As mentioned in #74, all the proposed changes for 0.10.0 can be done in a backwards compatible way, so maybe it would make sense to merge this as-is rather than making a judgement call one way or the other, because it clarifies how I'd expect base16 builders to react already.

**Comment by joshgoebel** [comment]
>  all the proposed changes for 0.10.0 can be done in a backwards compatible way,

I was referring to this issue in particular - the optional `#`... I did read MOST of that thread, but if there is something in there about this issue then I missed it.  In any case it wouldn't be super hard to update builders I suppose for such a change if needed. :)  Probably 1-2 LOC for the Node builder.

**Comment by belak** [comment]
Ah, my thought process went like this: the spec already says "should not" which doesn't explicitly forbid it like "must not" would, so logically we can assume builders should already handle the hash. Adding this language would just be clarifying what's already there.

**Comment by belak** [comment]
Closing in favor of #250 for now. There are no schemes I'm aware of currently doing this so it's not currently an issue that will even pop up.

**Comment by joshgoebel** [comment]
> Ah, my thought process went like this: the spec already says "should not" which doesn't explicitly forbid it like "must not" would, so logically we can assume builders should already handle the hash. Adding this language would just be clarifying what's already there.

That is _technically_ correct I suppose.

> so logically we can assume builders should already handle the hash

Just FYI: I just tested now and it entirely breaks the Node builder, since it's use `Buffer` to parse that data and expects it to be a pure hex string.  _I'm actually all for this change and the idea behind it now that I've thought about it..._ I just think that pretending that it wouldn't break anything or that all the builder writers are reading the spec as carefully as you are is a poor assumption. :-)  Honestly I personally would have read "should not" as "must not".

**Comment by belak** [comment]
> > Ah, my thought process went like this: the spec already says "should not" which doesn't explicitly forbid it like "must not" would, so logically we can assume builders should already handle the hash. Adding this language would just be clarifying what's already there.
> 
> That is _technically_ correct I suppose.

Yes, technically correct, the best form of correct. *insert dumb grin here*

> > so logically we can assume builders should already handle the hash
> 
> Just FYI: I just tested now and it entirely breaks the Node builder, since it's use `Buffer` to parse that data and expects it to be a pure hex string. _I'm actually all for this change and the idea behind it now that I've thought about it..._ I just think that pretending that it wouldn't break anything or that all the builder writers are reading the spec as carefully as you are is a poor assumption. :-)

That's a fair position - I didn't mean to pretend that if schemes were merged using that syntax, they wouldn't break existing builders.

> Honestly I personally would have read "should not" as "must not".

Most of my work is on the backend, and there if a protocol says "should not" you can bet some crazy implementation of a client somewhere does it, so I tend to code for those cases anyway, just in case. There's actually another line in the 0.10.0 spec about duplicate file-names I chose not to remove because *technically* if someone created a second scheme file matching another but with spaces in place of dashes in the primary repo, it could cause an issue. Yay edge cases I guess.
